### PR TITLE
Extract post metadata in new post_meta table

### PR DESCRIPTION
**Pull request by rshbhgrg**
**NOTE: The post metadata table split is purely an internal optimization for v3 and doesn't require or expect any external actions including related API usage in v3**

We keep running into issues adding new fields to the post table because there are too many fields making the post table "too wide". We have also hit MySQL limitations in how many bytes can be in a row (64kb) with post table.

For v3, the idea is to split the 8 post fields (meta, twitter and og) used for meta data into a `posts_meta` table as these 8 fields are all "problem" varchar fields and make sense logically grouped together. The API layer is unaffected by the split as input/output serializers ensure the data flow works the same way as it was in v2. This PR -

- [x] Creates new `post_meta` schema/table with 8 fields (2 `meta_* `,  3 `twitter_*` and 3 `og_*`) 
- [x] Update relations between `post` and `post_meta` table
- [x] Update input/output serializers to keep existing API behavior 
- [x] Avoids new entry in `post_meta` table for post where all meta fields are null
- [x] Updates `1:1` table relation handling in base model
- [x] Keeps the current fields API param behavior
- [x] Handles migration of existing posts to new table structure
  - [x] Adds new `post_meta` table
  - [x] Populates metadata entries for all posts which have non-null meta fields
  - [x] Removes 8 meta fields from `post` table
- [x] Updates tests
- [x] Updates importer/exporter to work seamlessly with table changes
- [ ] Makes the filter API parameters behave as intended (non-blocking edge-case, skipped for now)

### CODE REVIEW

**Comment by rishabhgrg** [comment]
@gargol @kevinansfield I am working on making tests/import/export work atm, but will be good to get eyes on major changes / migration in the meanwhile(when possible) 😬

File: core/server/data/schema/schema.js [hunk 1]
```diff
@@ -50,14 +48,20 @@ module.exports = {
         custom_excerpt: {type: 'string', maxlength: 2000, nullable: true, validations: {isLength: {max: 300}}},
         codeinjection_head: {type: 'text', maxlength: 65535, nullable: true},
         codeinjection_foot: {type: 'text', maxlength: 65535, nullable: true},
+        custom_template: {type: 'string', maxlength: 100, nullable: true},
+        canonical_url: {type: 'text', maxlength: 2000, nullable: true}
+    },
+    post_meta: {
```
[the remark this answers is not in the document]
Comment (gargol) [reply core/server/data/schema/schema.js]: how else would you call it? 
think `posts_meta` is fine as it has already been proposed and agreed upon in the spec :wink:

[the remark this answers is not in the document]
Comment (rishabhgrg) [reply core/server/data/schema/schema.js]: lol, yeah makes sense once I zoom back out, will update the naming, thanks :D

File: core/server/api/canary/utils/serializers/output/utils/mapper.js [hunk 2]
```diff
@@ -57,6 +58,17 @@ const mapPost = (model, frame) => {
         });
     }
 
+    // Transforms post/page metadata to flat structure
+    let postsMeta = Object.assign({}, _.mapValues(postsMetaSchema, () => null), jsonModel.posts_meta);
```
Comment (gargol) [inline core/server/api/canary/utils/serializers/output/utils/mapper.js]: For the sake of keeping the syntax consistent maybe worth doing similar transformation to the one in the input serializer:
```
let metaAttrs = _.keys(_.omit(postsMetaSchema, ['id', 'post_id']));
```

This would avoid doing `delete` statements afterward.

Comment (gargol) [reply core/server/api/canary/utils/serializers/output/utils/mapper.js]: In addition, `metaAttrs` could be filtered with `!frame.options.columns || (frame.options.columns && frame.options.columns.includes(k)))` so that later on it's just a flat assignment to jsonMode :wink:

File: core/server/data/importer/importers/data/posts.js [hunk 3]
```diff
@@ -4,13 +4,15 @@ const uuid = require('uuid');
...
+const postsMetaSchema = require('../../../schema').tables.posts_meta;
+const metaAttrs = _.keys(_.omit(postsMetaSchema, ['id']));
 
 class PostsImporter extends BaseImporter {
     constructor(allDataFromFile) {
         super(allDataFromFile, {
             modelName: 'Post',
             dataKeyToImport: 'posts',
-            requiredFromFile: ['posts', 'tags', 'posts_tags', 'posts_authors'],
+            requiredFromFile: ['posts', 'tags', 'posts_tags', 'posts_authors', 'posts_meta'],
```
Comment (gargol) [inline core/server/data/importer/importers/data/posts.js]: How does this behave when there are no records in `posts_meta` table? Just wondering about the situation where all posts don't have any records for `meta_title`, `meta_description` etc.

File: core/server/data/importer/importers/data/posts.js [hunk 4]
```diff
@@ -24,6 +26,18 @@ class PostsImporter extends BaseImporter {
         });
     }
 
+    /**
+     * Sanitizes post metadata, picking data from sepearate table(for >= v3) or post itself(for < v3)
+     */
+    santizePostsMeta(model) {
```
Comment (gargol) [inline core/server/data/importer/importers/data/posts.js]:
```suggestion
    sanitizePostsMeta(model) {
```

Small typo ;)

File: core/server/data/importer/importers/data/posts.js [hunk 5]
```diff
@@ -202,6 +216,7 @@ class PostsImporter extends BaseImporter {
                 model.mobiledoc = JSON.stringify(mobiledoc);
                 model.html = converters.mobiledocConverter.render(JSON.parse(model.mobiledoc));
             }
+            this.santizePostsMeta(model);
```
Comment (gargol) [inline core/server/data/importer/importers/data/posts.js]:
```suggestion
            this.sanitizePostsMeta(model);
```

same as above

File: core/server/data/schema/commands.js [hunk 6]
```diff
@@ -149,7 +149,7 @@ function createColumnMigration(...migrations) {
         const hasColumn = await conn.schema.hasColumn(table, column);
         const isInCorrectState = dbIsInCorrectState(hasColumn);
 
-        const log = createLog(isInCorrectState ? 'info' : 'warn');
+        const log = createLog(isInCorrectState ? 'warn' : 'info');
```
Comment (gargol) [inline core/server/data/schema/commands.js]: Why was this changed?

Comment (rishabhgrg) [reply core/server/data/schema/commands.js]: Was changed as this is the right expected behavior and was a bug in current code. We want to `warn` in migration when they try to run a migration even though its already in correct state, but give them info when its not and migration did its job 😄

File: core/server/models/post.js [hunk 7]
```diff
@@ -232,6 +233,18 @@ Post = ghostBookshelf.Model.extend({
                 authors.forEach(author => author.emitChange('attached', options));
             });
         });
+        
+        model.related('posts_meta').once('detaching', function onDetached(collection, postsMeta) {
+            model.related('posts_meta').once('detached', function onDetached(detachedCollection, response, options) {
+                postsMeta.emitChange('detached', options);
```
Comment (gargol) [inline core/server/models/post.js]: Do we have a use-case for this event? We are only emitting them for Webhooks and some special communication with the "frontend". Not sure it makes sense to have them unless there's a specific need? :thinking:

**Comment by gargol** [comment]
@rishabhgrg gave this PR the first pass and left some comments to be discussed :smiley: Think it's in pretty good shape. Just want to give it a bit more testing time with importer/exporter to make sure everything works as intended.

File: core/server/data/schema/commands.js (hunk 6 above)
Comment (gargol) [reply core/server/data/schema/commands.js]: Oh! Then maybe worth pushing this straight to master ;)

File: core/server/data/importer/importers/data/posts.js (hunk 3 above)
Comment (rishabhgrg) [reply core/server/data/importer/importers/data/posts.js]: Context: This reads value from imported json file and is used [here](https://github.com/TryGhost/Ghost/pull/11102/files#diff-e712df50c0dc7cf33746eeff0564003cR33). 

In case of v3 export, we will have `posts_meta: []` entry in the exported json, and in case of v2 exports, the `this.requiredFromFile.posts_meta` value will be `null`. In both cases, since we are using `_.find` from lodash, it will handle it and return `null` for `postsMetaFromFile`. 😄

File: core/server/models/post.js (hunk 7 above)
Comment (rishabhgrg) [reply core/server/models/post.js]: 🤔 Hmm yeah didn't have any usecase for them in mind, just added them for consistency with rest of relations. I'll remove this and the `emitChange` method added in `postsMeta` table 👍

File: core/server/data/schema/commands.js (hunk 6 above)
Comment (rishabhgrg) [reply core/server/data/schema/commands.js]: Sure, I'll do that as separate.

File: core/server/models/post.js (hunk 7 above)
Comment (rishabhgrg) [reply core/server/models/post.js]: @gargol Done 🔥

File: core/server/data/schema/commands.js (hunk 6 above)
Comment (rishabhgrg) [reply core/server/data/schema/commands.js]: @gargol Also Done 🔥

File: core/server/data/importer/importers/data/posts.js (hunk 3 above)
Comment (gargol) [reply core/server/data/importer/importers/data/posts.js]: Awesome! No further Qs on this one :)

File: core/server/api/canary/utils/serializers/output/utils/mapper.js (hunk 2 above)
Comment (rishabhgrg) [reply core/server/api/canary/utils/serializers/output/utils/mapper.js]: @gargol Thanks, good suggestion, have cleaned up the code in mappers and it looks much cleaner now. lmk if this looks good :)
